ethapi/api.go

```diff
@@ -710,8 +709,14 @@ type StorageResult struct {
 
 // GetProof returns the Merkle-proof for a given account and optionally some storage keys.
 func (s *PublicBlockChainAPI) GetProof(ctx context.Context, address common.Address, storageKeys []string, blockNrOrHash rpc.BlockNumberOrHash) (*AccountResult, error) {
-	state, _, err := s.b.StateAndHeaderByNumberOrHash(ctx, blockNrOrHash)
-	if state == nil || err != nil {
+	statedb, _, err := s.b.StateAndHeaderByNumberOrHash(ctx, blockNrOrHash)
+	if statedb == nil || err != nil {
+		return nil, err
+	}
+
+	state, ok := evmcore.IsMptStateDB(statedb)
```

Comment:
Sorry for asking, was just looking into the code and the consensus and found PRs.
How do you think, does it worth it to use generics+interfaces here to get rid of type assertion and have more strict type checks?

Reply:
It will be an over-engineering for now. All we need is check here does the proofs complies with the legacy format.
We should not return others kind of proofs because clients understand the legacy (MPT) only.